File: MTA_Hacks_2026_ios/SealSensei/Services/GradeStore.swift
//
//  GradeStore.swift
//  Studious Seal
//
//  Persists folders, classes, and grade components to UserDefaults.
//

import Foundation
import SwiftUI

// MARK: - Persisted data

private struct GradeData: Codable {
    var folders: [GradeFolder]
    var classes: [GradeClass]
    var components: [GradeComponent]
    var displayMode: GradeDisplayMode
}

// MARK: - Store

@MainActor
final class GradeStore: ObservableObject {
    static let displayModeKey = "sealsensei.gradeDisplayMode"
    static let gradeDataKey = "sealsensei.gradeData"

    @Published var folders: [GradeFolder] = []
    @Published var classes: [GradeClass] = []
    @Published var components: [GradeComponent] = []
    @Published var displayMode: GradeDisplayMode = .percentage

    private let defaults = UserDefaults.standard

    init() {
        load()
    }

    // MARK: - Load / Save

    private func load() {
        if let raw = defaults.string(forKey: Self.gradeDataKey),
           let data = try? JSONDecoder().decode(GradeData.self, from: Data(raw.utf8)) {
            folders = data.folders
            classes = data.classes
            components = data.components
            displayMode = data.displayMode
        } else if let modeRaw = defaults.string(forKey: Self.displayModeKey),
                  let mode = GradeDisplayMode(rawValue: modeRaw) {
            displayMode = mode
        }
    }

    private func save() {
        let data = GradeData(folders: folders, classes: classes, components: components, displayMode: displayMode)
        if let encoded = try? JSONEncoder().encode(data),
           let string = String(data: encoded, encoding: .utf8) {
            defaults.set(string, forKey: Self.gradeDataKey)
        }
        defaults.set(displayMode.rawValue, forKey: Self.displayModeKey)
    }

    // MARK: - Folders

    func addFolder(name: String) {
        folders.append(GradeFolder(name: name))
        folders.sort { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
        save()
    }

    func updateFolder(_ folder: GradeFolder) {
        if let i = folders.firstIndex(where: { $0.id == folder.id }) {
            folders[i] = folder
            save()
        }
    }

    func deleteFolder(_ folder: GradeFolder) {
        let classIdsInFolder = Set(classes.filter { $0.folderId == folder.id }.map(\.id))
        components.removeAll { classIdsInFolder.contains($0.classId) }
        classes.removeAll { $0.folderId == folder.id }
        folders.removeAll { $0.id == folder.id }
        save()
    }

    // MARK: - Classes

    func classes(in folderId: UUID) -> [GradeClass] {
        classes.filter { $0.folderId == folderId }.sorted { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
    }

    func unfiledClasses() -> [GradeClass] {
        classes.filter { $0.folderId == nil }.sorted { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
    }

    func addClass(name: String, folderId: UUID?) {
        classes.append(GradeClass(name: name, folderId: folderId))
        save()
    }

    func updateClass(_ gradeClass: GradeClass) {
        if let i = classes.firstIndex(where: { $0.id == gradeClass.id }) {
            classes[i] = gradeClass
            save()
        }
    }

    func deleteClass(_ gradeClass: GradeClass) {
        components.removeAll { $0.classId == gradeClass.id }
        classes.removeAll { $0.id == gradeClass.id }
        save()
    }

    /// Move a class into a folder (or back to unfiled if folderId is nil).
    func moveClass(_ gradeClass: GradeClass, toFolderId folderId: UUID?) {
        var updated = gradeClass
        updated.folderId = folderId
        updateClass(updated)
    }

    // MARK: - Components

    func components(for classId: UUID) -> [GradeComponent] {
        components.filter { $0.classId == classId }.sorted { $0.name.localizedCaseInsensitiveCompare($1.name) == .orderedAscending }
    }

    func addComponent(classId: UUID, name: String, weightPercent: Double, earnedPoints: Double, maxPoints: Double) {
        components.append(GradeComponent(classId: classId, name: name, weightPercent: weightPercent, earnedPoints: earnedPoints, maxPoints: maxPoints))
        save()
    }

    func updateComponent(_ component: GradeComponent) {
        if let i = components.firstIndex(where: { $0.id == component.id }) {
            components[i] = component
            save()
        }
    }

    func deleteComponent(_ component: GradeComponent) {
        components.removeAll { $0.id == component.id }
        save()
    }

    /// Overall grade for a class (weighted average of components). 0–100.
    func overallPercentage(for classId: UUID) -> Double? {
        let comps = components(for: classId)
        guard !comps.isEmpty else { return nil }
        let totalWeight = comps.reduce(0) { $0 + $1.weightPercent }
        guard totalWeight > 0 else { return nil }
        let weightedSum = comps.reduce(0) { $0 + $1.weightedContribution }
        return weightedSum / totalWeight * 100
    }

    func setDisplayMode(_ mode: GradeDisplayMode) {
        displayMode = mode
        save()
    }
}

File: MTA_Hacks_2026_Web/src/services/gradeStore.ts
import type { GradeDisplayMode, GradeFolder, GradeClass, GradeComponent } from '../types/grades'
import { componentPercentage, weightedContribution } from '../types/grades'

const GRADE_DATA_KEY = 'sealsensei.gradeData'

interface PersistedGradeData {
  folders: GradeFolder[]
  classes: GradeClass[]
  components: GradeComponent[]
  display_mode: GradeDisplayMode
}

function uuid(): string {
  return crypto.randomUUID()
}

function load(): PersistedGradeData {
  try {
    const raw = localStorage.getItem(GRADE_DATA_KEY)
    if (!raw) return defaultData()
    const data = JSON.parse(raw) as PersistedGradeData
    return {
      folders: data.folders ?? [],
      classes: data.classes ?? [],
      components: data.components ?? [],
      display_mode: data.display_mode === 'points' ? 'points' : 'percentage',
    }
  } catch {
    return defaultData()
  }
}

function defaultData(): PersistedGradeData {
  return {
    folders: [],
    classes: [],
    components: [],
    display_mode: 'percentage',
  }
}

function save(data: PersistedGradeData) {
  try {
    localStorage.setItem(GRADE_DATA_KEY, JSON.stringify(data))
  } catch {}
}

export interface GradeStoreState {
  folders: GradeFolder[]
  classes: GradeClass[]
  components: GradeComponent[]
  displayMode: GradeDisplayMode
}

export function createGradeStore(): GradeStoreState {
  const data = load()
  return {
    folders: data.folders,
    classes: data.classes,
    components: data.components,
    displayMode: data.display_mode,
  }
}

function toPersisted(state: GradeStoreState): PersistedGradeData {
  return {
    folders: state.folders,
    classes: state.classes,
    components: state.components,
    display_mode: state.displayMode,
  }
}

export function addFolder(state: GradeStoreState): GradeStoreState {
  const name = prompt('Folder name')?.trim()
  if (!name) return state
  const next: GradeStoreState = {
    ...state,
    folders: [...state.folders, { id: uuid(), name, created_at: new Date().toISOString() }].sort((a, b) => a.name.localeCompare(b.name, undefined, { sensitivity: 'base' })),
  }
  save(toPersisted(next))
  return next
}

export function deleteFolder(state: GradeStoreState, folder: GradeFolder): GradeStoreState {
  const classIds = new Set(state.classes.filter((c) => c.folder_id === folder.id).map((c) => c.id))
  const next: GradeStoreState = {
    ...state,
    folders: state.folders.filter((f) => f.id !== folder.id),
    classes: state.classes.filter((c) => c.folder_id !== folder.id),
    components: state.components.filter((c) => !classIds.has(c.class_id)),
  }
  save(toPersisted(next))
  return next
}

export function classesInFolder(state: GradeStoreState, folderId: string): GradeClass[] {
  return state.classes.filter((c) => c.folder_id === folderId).sort((a, b) => a.name.localeCompare(b.name, undefined, { sensitivity: 'base' }))
}

export function unfiledClasses(state: GradeStoreState): GradeClass[] {
  return state.classes.filter((c) => c.folder_id == null).sort((a, b) => a.name.localeCompare(b.name, undefined, { sensitivity: 'base' }))
}

export function addClass(state: GradeStoreState, name: string, folderId: string | null): GradeStoreState {
  const next: GradeStoreState = {
    ...state,
    classes: [...state.classes, { id: uuid(), name, folder_id: folderId, created_at: new Date().toISOString() }].sort((a, b) => a.name.localeCompare(b.name, undefined, { sensitivity: 'base' })),
  }
  save(toPersisted(next))
  return next
}

export function deleteClass(state: GradeStoreState, gradeClass: GradeClass): GradeStoreState {
  const next: GradeStoreState = {
    ...state,
    classes: state.classes.filter((c) => c.id !== gradeClass.id),
    components: state.components.filter((c) => c.class_id !== gradeClass.id),
  }
  save(toPersisted(next))
  return next
}

export function moveClass(state: GradeStoreState, gradeClass: GradeClass, toFolderId: string | null): GradeStoreState {
  const next: GradeStoreState = {
    ...state,
    classes: state.classes.map((c) => (c.id === gradeClass.id ? { ...c, folder_id: toFolderId } : c)),
  }
  save(toPersisted(next))
  return next
}

export function componentsForClass(state: GradeStoreState, classId: string): GradeComponent[] {
  return state.components.filter((c) => c.class_id === classId).sort((a, b) => a.name.localeCompare(b.name, undefined, { sensitivity: 'base' }))
}

export function overallPercentage(state: GradeStoreState, classId: string): number | null {
  const comps = componentsForClass(state, classId)
  if (comps.length === 0) return null
  const totalWeight = comps.reduce((s, c) => s + c.weight_percent, 0)
  if (totalWeight <= 0) return null
  const sum = comps.reduce((s, c) => s + weightedContribution(c), 0)
  return (sum / totalWeight) * 100
}

export function addComponent(state: GradeStoreState, classId: string, name: string, weightPercent: number, earnedPoints: number, maxPoints: number): GradeStoreState {
  const next: GradeStoreState = {
    ...state,
    components: [...state.components, { id: uuid(), class_id: classId, name, weight_percent: weightPercent, earned_points: earnedPoints, max_points: maxPoints }],
  }
  save(toPersisted(next))
  return next
}

export function updateComponent(state: GradeStoreState, component: GradeComponent): GradeStoreState {
  const next: GradeStoreState = {
    ...state,
    components: state.components.map((c) => (c.id === component.id ? component : c)),
  }
  save(toPersisted(next))
  return next
}

export function deleteComponent(state: GradeStoreState, component: GradeComponent): GradeStoreState {
  const next: GradeStoreState = {
    ...state,
    components: state.components.filter((c) => c.id !== component.id),
  }
  save(toPersisted(next))
  return next
}

export function setDisplayMode(state: GradeStoreState, mode: GradeDisplayMode): GradeStoreState {
  const next = { ...state, displayMode: mode }
  save(toPersisted(next))
  return next
}

export function formatGradeDisplay(store: GradeStoreState, pct: number): string {
  return store.displayMode === 'percentage' ? `${pct.toFixed(1)}%` : `${pct.toFixed(1)} / 100`
}

export function formatComponentDisplay(store: GradeStoreState, c: GradeComponent): string {
  const pct = componentPercentage(c)
  return store.displayMode === 'percentage' ? `${pct.toFixed(1)}%` : `${c.earned_points} / ${c.max_points}`
}
